pkg/controllers/vdb/subclustershutdown_reconciler.go

```diff
@@ -0,0 +1,160 @@
+	VRec       config.ReconcilerInterface
+	Log        logr.Logger
+	Vdb        *vapi.VerticaDB // Vdb is the CRD we are acting on
+	Dispatcher vadmin.Dispatcher
+	PFacts     *podfacts.PodFacts
+}
+
+// MakeSubclusterShutdownReconciler will build a SubclusterShutdownReconciler object
+func MakeSubclusterShutdownReconciler(recon config.ReconcilerInterface, log logr.Logger,
+	vdb *vapi.VerticaDB, dispatcher vadmin.Dispatcher, pfacts *podfacts.PodFacts) controllers.ReconcileActor {
+	return &SubclusterShutdownReconciler{
+		VRec:       recon,
+		Log:        log.WithName("SubclusterShutdownReconciler"),
+		Vdb:        vdb,
+		Dispatcher: dispatcher,
+		PFacts:     pfacts,
+	}
+}
+
+func (s *SubclusterShutdownReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	if err := s.PFacts.Collect(ctx, s.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+	subclusters, err := s.getSubclustersToShutdown()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	for scName, initIP := range subclusters {
+		err := s.PFacts.RemoveStartupFileInSubclusterPods(ctx, scName, "removed startup.json before stop_subcluster")
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		err = s.runStopSubclusterVclusterAPI(ctx, scName, initIP)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		s.PFacts.Invalidate()
+	}
+	return ctrl.Result{}, nil
+}
+
+// getSubclustersToShutdown returns the subclusters that need to get
+// shut down
+func (s *SubclusterShutdownReconciler) getSubclustersToShutdown() (map[string]string, error) {
+	subclusters := map[string]string{}
+	primarySubclusters := []string{}
+	upPrimaryNodes := 0
+	willLoseQuorum := false
+	scSbMap := s.Vdb.GenSubclusterSandboxMap()
+	sbMap := s.Vdb.GenSandboxMap()
+	for i := range s.Vdb.Spec.Subclusters {
+		sc := &s.Vdb.Spec.Subclusters[i]
+		sandbox := scSbMap[sc.Name]
+		if sandbox != s.PFacts.GetSandboxName() {
+			continue
+		}
+		// no-op if the subcluster is not marked for
+		// shutdown
+		if !sc.Shutdown {
+			continue
+		}
+		if sandbox != vapi.MainCluster {
+			sb := sbMap[sandbox]
+			// no-op if the subcluster shutdown is driven
+			// by its sandbox
+			if sb != nil && sb.Shutdown {
+				continue
+			}
+		}
+		hostIP, ok := s.PFacts.FindFirstUpPodIP(false, sc.Name)
+		if !ok {
+			continue
+		}
+		if sc.IsPrimary() {
+			primarySubclusters = append(primarySubclusters, sc.Name)
+			upPrimaryNodes += s.PFacts.GetSubclusterUpNodeCount(sc.Name)
+			// If stopping a subcluster would cause cluster quorum, we abort
+			// the operation
+			if !s.PFacts.DoesDBHaveQuorum(upPrimaryNodes) {
+				willLoseQuorum = true
+				break
+			}
+		}
+		subclusters[sc.Name] = hostIP
+	}
+	if willLoseQuorum {
+		s.VRec.Eventf(s.Vdb, corev1.EventTypeWarning, events.ClusterWillLoseQuorum,
+			"Shutting down subclusters %s will cause quorum loss.", strings.Join(primarySubclusters, ","))
+		return subclusters, fmt.Errorf("cannot shut down primaries %s because it will cause quorum, loss. "+
```

Comment:
I will. But We need to be careful with this one. I have a few concerns:
-  Once the cluster is read-only, the only way to bring it back is start_db. For that, all subclusters that have shutdown true, must have it set to false
- How about accidental shutdown. A case where I want to shutdown sec1(secondary), but I instead shutdown main(primary) (unlikely but still). Now my cluster is read-only against my will. I now need to start_db which means the db will be offline before going back online. Isn't very risky in production?
- Can a subcluster be stopped if the db is read-only(I think it cannot)? Which means the moment the db is read-only, you no longer can stop any subcluster.

I agree we need to support quorum-loss case, we should just think about it carefully that is why I prefer to exclude it for now. I always worry about that one user that one user that will make their cluster read-only by accident

Reply:
Okay, I see your concerns. If we need to support stop db/primay-scs in future, we could do a spike first.

I think start_db can start a subcluster instead of a database. So if the user accidentally shutdown the main, they should be able to start the main without stopping other subclusters. In future, we can use start_subcluster to better handle this case. And we have the draining shutdown which can reduce the risk.

For the question "Can a subcluster be stopped if the db is read-only", we need to test it in the spike.